File: app/src/main/java/cn/ntit/crps_re0/model/DynamicData.java
package cn.ntit.crps_re0.model;

import com.google.gson.annotations.SerializedName;

public class DynamicData {
    @SerializedName("vin") public float vin;
    @SerializedName("iin") public float iin;
    @SerializedName("vout") public float vout;
    @SerializedName("iout") public float iout;
    @SerializedName("pout") public float pout;
    @SerializedName("eff") public int eff;
    @SerializedName("fan") public int fan;
    @SerializedName("t1") public float t1;
    @SerializedName("t2") public float t2;
    @SerializedName("rt") public long rt;
    @SerializedName("pwr") public int pwr;
    @SerializedName("si") public int si;
    @SerializedName("sn") public int sn;
    @SerializedName("st") public int st;
    @SerializedName("sf") public int sf;

    public float getPin() {
        return vin * iin;
    }

    public boolean isPowerOn() {
        return pwr == 1;
    }
}

File: app/src/main/java/cn/ntit/crps_compose/data/DynamicData.kt
package cn.ntit.crps_compose.data

import org.json.JSONObject

/**
 * 动态遥测数据：ESP32 每 250ms 通过 BLE Notify 上报的 JSON。
 */
data class DynamicData(
    val vin: Float,
    val iin: Float,
    val vout: Float,
    val iout: Float,
    val pout: Float,
    val eff: Int,
    val fan: Int,
    val t1: Float,
    val t2: Float,
    val rt: Long,
    val pwr: Int,
    val si: Int,
    val sn: Int,
    val st: Int,
    val sf: Int,
) {
    /** 输入功率（计算值） */
    val pin: Float get() = vin * iin

    val isPowerOn: Boolean get() = pwr == 1

    companion object {
        fun parse(json: JSONObject): DynamicData = DynamicData(
            vin = json.optDouble("vin").toFloat(),
            iin = json.optDouble("iin").toFloat(),
            vout = json.optDouble("vout").toFloat(),
            iout = json.optDouble("iout").toFloat(),
            pout = json.optDouble("pout").toFloat(),
            eff = json.optInt("eff"),
            fan = json.optInt("fan"),
            t1 = json.optDouble("t1").toFloat(),
            t2 = json.optDouble("t2").toFloat(),
            rt = json.optLong("rt"),
            pwr = json.optInt("pwr"),
            si = json.optInt("si"),
            sn = json.optInt("sn"),
            st = json.optInt("st"),
            sf = json.optInt("sf"),
        )
    }
}
